homeassistant/components/remote/xiaomi_miio.py

```diff
@@ -0,0 +1,255 @@
+"""
+Support for the Xiaomi IR Remote (Chuangmi IR).
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/remote.xiaomi_miio/
+"""
+import asyncio
+import logging
+import time
+
+from datetime import timedelta
+
+import voluptuous as vol
+
+from homeassistant.components.remote import (
+    PLATFORM_SCHEMA, DOMAIN, ATTR_NUM_REPEATS, ATTR_DELAY_SECS,
+    DEFAULT_DELAY_SECS, RemoteDevice)
+from homeassistant.const import (
+    CONF_NAME, CONF_HOST, CONF_TOKEN, CONF_TIMEOUT,
+    ATTR_ENTITY_ID, ATTR_HIDDEN, CONF_COMMAND)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util.dt import utcnow
+
+REQUIREMENTS = ['python-miio==0.3.5']
+
+_LOGGER = logging.getLogger(__name__)
+
+SERVICE_LEARN = 'xiaomi_miio_learn_command'
+PLATFORM = 'xiaomi_miio'
+
+CONF_SLOT = 'slot'
+CONF_COMMANDS = 'commands'
+
+DEFAULT_TIMEOUT = 10
+DEFAULT_SLOT = 1
+
+LEARN_COMMAND_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): vol.All(str),
+    vol.Optional(CONF_TIMEOUT, default=10):
+        vol.All(int, vol.Range(min=0)),
+    vol.Optional(CONF_SLOT, default=1):
+        vol.All(int, vol.Range(min=1, max=1000000)),
+})
+
+COMMAND_SCHEMA = vol.Schema({
+    vol.Required(CONF_COMMAND): vol.All(cv.ensure_list, [cv.string])
+    })
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT):
+        vol.All(int, vol.Range(min=0)),
+    vol.Optional(CONF_SLOT, default=DEFAULT_SLOT):
+        vol.All(int, vol.Range(min=1, max=1000000)),
+    vol.Optional(ATTR_HIDDEN, default=True): cv.boolean,
+    vol.Required(CONF_TOKEN): vol.All(str, vol.Length(min=32, max=32)),
+    vol.Optional(CONF_COMMANDS, default={}):
+        vol.Schema({cv.slug: COMMAND_SCHEMA}),
+}, extra=vol.ALLOW_EXTRA)
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Xiaomi IR Remote (Chuangmi IR) platform."""
+    from miio import ChuangmiIr, DeviceException
+
+    host = config.get(CONF_HOST)
+    token = config.get(CONF_TOKEN)
+
+    # Create handler
+    _LOGGER.info("Initializing with host %s (token %s...)", host, token[:5])
+    device = ChuangmiIr(host, token)
+
+    # Check that we can communicate with device.
+    try:
+        device.info()
+    except DeviceException as ex:
+        _LOGGER.error("Token not accepted by device : %s", ex)
+        return
+
+    if PLATFORM not in hass.data:
+        hass.data[PLATFORM] = {}
```

Comment:
Maybe you should use a key specific for the remote xiaomi_miio platform? Other xiaomi_miio platforms might want to store things in hass.data in the future.

Reply:
I am pretty sure that this part is correct, if you look at the other python-miio platforms they do the same thing: Create a dict if one is not on the PLATFORM key, and save the devices in the new dict using their host (ip) as key. The host _should_ be unique but there may be some edge cases where it is not.

See these lines in the vacuum python-miio:
https://github.com/home-assistant/home-assistant/blob/5ba02c531e4d7e68dc8b811ba2ebc7438afe6f16/homeassistant/components/vacuum/xiaomi_miio.py#L91-L103